presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java

```diff
@@ -197,20 +201,37 @@ public TaskStatus process()
 
     private void invokeNoMoreSplitsIfNecessary()
     {
-        if (partitions.isEmpty() && fileIterators.isEmpty()) {
-            taskExecutionLock.writeLock().lock();
-            try {
-                // the write lock guarantees that no one is operating on the partitions, fileIterators, or hiveSplitSource, or half way through doing so.
-                if (partitions.isEmpty() && fileIterators.isEmpty()) {
-                    // It is legal to call `noMoreSplits` multiple times or after `stop` was called.
-                    // Nothing bad will happen if `noMoreSplits` implementation calls methods that will try to obtain a read lock because the lock is re-entrant.
-                    hiveSplitSource.noMoreSplits();
-                }
+        taskExecutionLock.readLock().lock();
+        try {
+            if (!partitions.isEmpty() || !fileIterators.isEmpty()) {
```

Comment:
Comment that this is an opportunistic check to avoid trying to acquire the write lock necessarily.

Reply:
Updated.